MaxText/layers/models.py

```diff
@@ -819,3 +817,72 @@ def __call__(
         image_embeddings=image_embeddings,
     )
     return logits
+
+class ZeroOneTransformer(nn.Module):
+  """
+  A wrapper for the base Transformer model designed to implement the Zero-1
+  FSDP optimization.
+
+  The goal of this optimization is to reduce communication overhead. In the standard
+  FSDP implementation, an all-gather operation on the model weights is performed twice
+  for each gradient accumulation microbatch (once for the forward pass, once for the backward pass).
+  This class changes that behavior. When enabled, it performs the all-gather operation
+  only *once* per full gradient accumulation step. It gathers the full weights into
+  memory, runs all the microbatch forward and backward passes, and then releases the
+  full weights. This trades higher peak memory usage for significantly reduced
+  network communication, which can improve training speed if sufficient memory is
+  available.
+  """
+
+  config: Config
+  mesh: Mesh
+  quant: Quant
+
+  def setup(self):
+    self.model = Transformer(self.config, self.mesh, self.quant)
+
+  def __call__(
+      self,
+      decoder_input_tokens: jnp.ndarray,
+      decoder_positions: jnp.ndarray,
+      decoder_segment_ids=None,
+      encoder_images: Optional[jnp.ndarray]=None,
+      enable_dropout=True,
+      model_mode=MODEL_MODE_TRAIN,
+      previous_chunk=None,
+      true_length: Optional[int] = None,
+      slot: Optional[int] = None,
+      page_state: Optional[page_manager.PageState] = None,
+      partition_spec=None,
+  ):
+    if self.is_initializing():
+      return self.model(
+          decoder_input_tokens,
+          decoder_positions,
+          decoder_segment_ids,
+          encoder_images,
+          enable_dropout,
+          model_mode,
+          previous_chunk,
+          true_length,
+          slot,
+          page_state,
```

Comment:
Nit: Ideally would use the named params for passing these in. Really minor though